google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -350,6 +355,24 @@ def _auth_init(self) -> None:
 
         self._credentials = credentials
 
+    async def _force_refresh(self) -> bool:
+        try:
+            self._current = await self._perform_refresh()
+            return True
+        except asyncio.queues.QueueFull:
+            # a refresh attempt is already queued, so just block on the result
+            self._current = await self._next
+            return True
+        except Exception as e:
```

Comment:
This logic seems wrong when compared to the java connector: https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java#L307-L313

I think we probably want to mirror that behavior here?

Reply:
The code for python is going to look a little bit different because Python tasks don't behave like Java futures. 

https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel
Java futures [can reject cancellation if a task is in progress](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#cancel(boolean)). When an asyncio task is cancelled, it throws a `CancelledError` in the wrapped coroutine. The coroutine can then either handle and suppress or throw that error. Since we're just wrapping `_perform_refresh` in a coroutine with a sleep (`_schedule_refresh`) to schedule the next refresh instead of an executor like in Java, the refresh will just get cancelled, regardless of whether it's sleeping or running.

The code above attempts another refresh, but if the rate limiter queue is full, we know there's a `_perform_refresh` call that's about to run. If schedule_refresh is still sleeping, then the queue will be empty. If the queue is full, we just block on the pending `_next` result, like in Java. We don't have to update `_next` ourselves, because awaiting `_perform_refresh()` will schedule the next task.